include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,401 @@
+    const double            dq_relative_error,
+    const unsigned int      maximum_beta_failures,
+    const unsigned int      anderson_subspace_size,
+    const double            absolute_tolerance,
+    const double            relative_tolerance,
+    const unsigned int      threshold_nonlinear_iterations,
+    const unsigned int      threshold_n_linear_iterations,
+    const bool              reuse_solver)
+    : solver_type(solver_type)
+    , strategy(strategy)
+    , maximum_non_linear_iterations(maximum_non_linear_iterations)
+    , function_tolerance(function_tolerance)
+    , step_tolerance(step_tolerance)
+    , no_init_setup(no_init_setup)
+    , maximum_setup_calls(maximum_setup_calls)
+    , maximum_newton_step(maximum_newton_step)
+    , dq_relative_error(dq_relative_error)
+    , maximum_beta_failures(maximum_beta_failures)
+    , anderson_subspace_size(anderson_subspace_size)
+    , absolute_tolerance(absolute_tolerance)
+    , relative_tolerance(relative_tolerance)
+    , threshold_nonlinear_iterations(threshold_nonlinear_iterations)
+    , threshold_n_linear_iterations(threshold_n_linear_iterations)
+    , reuse_solver(reuse_solver)
+  {}
+
+// IDEAS:
+/**
+ * KINSOL has "AdditionalData"
+ * NOX has "AdditionalData" and "Parameters"
+ * 
+ * Make AdditionalData class for NLSolverSelector that distributes
+ * to both AdditionalData classes accordingly.
+ * 
+ * Include Parameters class for NOX specifically.
+ * 
+ * Have checks in the solve_NL() function that changes the solver
+ * type depending on the specifications, i.e., Line search vs Picard
+*/
+
+template <typename VectorType>
+void
+NLSolverSelector<VectorType>::set_data(
+  const typename TrilinosWrappers::NOXSolver<VectorType>::AdditionalData &additional_data,
+  const Teuchos::RCP<Teuchos::ParameterList> &parameters) 
+{
+  #ifdef DEAL_II_WITH_TRILINOS
+  additional_data_nox = additional_data;
+  parameters_nox = parameters;
+  #endif
+}
+
+template <typename VectorType>
+void
+NLSolverSelector<VectorType>::set_data
+  (const typename SUNDIALS::KINSOL<VectorType>::AdditionalData &additional_data)
+{
+  additional_data_kinsol = additional_data;
+}
+
+template <typename VectorType>
+void
+NLSolverSelector<VectorType>::solve(VectorType &initial_guess_and_solution)
+{
+  if (additional_data.solver_type == "kinsol")
+    {
+      std::cout << "Maximum nonlinear iterations: "
+                << additional_data_kinsol.maximum_non_linear_iterations
+                << std::endl;
+      std::cout << std::endl;
+
+      SUNDIALS::KINSOL<VectorType> nonlinear_solver(additional_data_kinsol);
+
+      nonlinear_solver.reinit_vector = 
+        [&](Vector<double> &x) {
+          NLSolverSelector<VectorType>::reinit_vector(x);
+        };
+
+      nonlinear_solver.residual = 
+        [&](const Vector<double> &evaluation_point,
+            Vector<double> &      residual) {
+          NLSolverSelector<VectorType>::residual(evaluation_point, residual);
+    
+          return 0;
```

Comment:
I think you should return the return value of residual(). Here and in all other functions.

Reply:
I'm a bit confused here, all the examples it's always been void functions and return 0. What do you mean by residual()? That's a vector right?

Reply:
do ``return NLSolverSelector<VectorType>::residual(evaluation_point, residual);``